src/common/walFilter/versions/recordProcessGPDB6.c

```diff
@@ -0,0 +1,503 @@
+#include "build.auto.h"
+
+#include <stdbool.h>
+#include "common/log.h"
+#include "postgres/interface/crc32.h"
+#include "recordProcessGPDB6.h"
+#include "xlogInfoGPDB6.h"
+
+#define XLR_INFO_MASK           0x0F
+#define XLOG_HEAP_OPMASK        0x70
+
+typedef uint32 CommandId;
+typedef uint16 OffsetNumber;
+typedef uint32 ForkNumber;
+
+typedef struct BkpBlock
+{
+    RelFileNode node;           /* relation containing block */
+    ForkNumber fork;            /* fork within the relation */
+    BlockNumber block;          /* block number */
+    uint16 hole_offset;         /* number of bytes before "hole" */
+    uint16 hole_length;         /* number of bytes in "hole" */
+
+    /* ACTUAL BLOCK DATA FOLLOWS AT END OF STRUCT */
+} BkpBlock;
+
+typedef struct xl_smgr_truncate
+{
+    BlockNumber blkno;
+    RelFileNode rnode;
+} xl_smgr_truncate;
+
+typedef struct BlockIdData
+{
+    uint16 bi_hi;
+    uint16 bi_lo;
+} BlockIdData;
+
+typedef struct ItemPointerData
+{
+    BlockIdData ip_blkid;
+    OffsetNumber ip_posid;
+}
+
+#ifdef __arm__
+__attribute__((packed))         /* Appropriate whack upside the head for ARM */
+#endif
+ItemPointerData;
+
+typedef struct xl_heaptid
+{
+    RelFileNode node;
+    ItemPointerData tid;        /* changed tuple id */
+} xl_heaptid;
+
+typedef struct xl_heap_new_cid
+{
+    /*
+     * store toplevel xid so we don't have to merge cids from different
+     * transactions
+     */
+    TransactionId top_xid;
+    CommandId cmin;
+    CommandId cmax;
+
+    /*
+     * don't really need the combocid since we have the actual values right in
+     * this struct, but the padding makes it free and its useful for
+     * debugging.
+     */
+    CommandId combocid;
+
+    /*
+     * Store the relfilenode/ctid pair to facilitate lookups.
+     */
+    xl_heaptid target;
+} xl_heap_new_cid;
+
+// Get RelFileNode from XLOG record.
+// Only XLOG_FPI contains RelFileNode, so the other record types are ignored.
+static bool
+getXlog(const XLogRecord *record, RelFileNode **node)
```

Comment:
Why not make the node a return argument for these functions, returning null when ignored, like this?
```suggestion
static RelFileNode *
getXlog(const XLogRecord *record)
```

Reply:
done